File: 백준/Silver/12867. N차원 여행/N차원 여행.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
//        BufferedReader br = new BufferedReader(new FileReader("./input.txt"));

        int n = Integer.parseInt(br.readLine());
        int m = Integer.parseInt(br.readLine());
        int[] indexOrder = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();
        char[] dir = br.readLine().toCharArray();

        Map<Integer, Integer> cur = new HashMap<>();
        Set<Map<Integer, Integer>> visit = new HashSet<>();
        visit.add(new HashMap<>(cur));

        for (int i = 0; i < m; i++) {
            int idx = indexOrder[i];
            int change = dir[i] == '+' ? 1 : -1;
            cur.put(idx, cur.getOrDefault(idx, 0) + change);

            if (cur.get(idx) == 0) {
                cur.remove(idx);
            }

            if (!checkVisit(visit, cur)) {
                System.out.println(0);
                return;
            }
            visit.add(new HashMap<>(cur));
        }
        System.out.println(1);
    }

    public static boolean checkVisit(Set<Map<Integer, Integer>> visit, Map<Integer, Integer> cur) {
        for (Map<Integer, Integer> visitPoint : visit) {
            if (visitPoint.equals(cur)) {
                return false;
            }
        }
        return true;
    }
}

File: 백준/Silver/12867. N차원 여행/N차원 여행.py
N = int(input())
M = int(input())
dims = list(map(int, input().split()))
dir = input()

able = True
path = [{}]
cur = {}

for i in range(M):
    cur[dims[i]] = cur.get(dims[i], 0) + (1 if dir[i] == '+' else -1)
    if cur[dims[i]] == 0:               # 위치가 0 인 차원은 딕셔너리에서 제거함으로써 모든 차원 표현할 필요 없음)
        del cur[dims[i]]
    for cmp in path:
        if cur.items() == cmp.items():  # 지나왔던 경로에 현재 위치가 존재하면 불가능
            able = False
            break
    if not able:
        break
    path.append(cur.copy())             # 경로에 현재 위치 추가

if cur == {}: able = False      # 마지막이 원점일 경우에도 불가능
print(1 if able else 0)

